source/tutorials/nixos-configuration-on-vm.md

```diff
@@ -0,0 +1,169 @@
+You can create a NixOS configuration in your working directory:
+```shell-session
+nixos-generate-config --dir ./
+```
+
+In the working directory you will find two files: `configuration.nix` and `hardware-configuration.nix`.
+
+The `hardware-configuration.nix` file is specific to the hardware the `nix-generate-config` is being run on.
+We can ignore that file for this tutorial because it has no effect inside a virtual machine.
+
+The `configuration.nix` file contains various suggestions for the initial setup of a desktop computer.
+Without comments the configuration file contains the following content:
+```nix
+{ config, pkgs, ... }:
+{
+  imports =
+    [
+      ./hardware-configuration.nix
+    ];
+
+  boot.loader.systemd-boot.enable = true;
+  boot.loader.efi.canTouchEfiVariables = true;
+
+  services.xserver.enable = true;
+
+  services.xserver.displayManager.gdm.enable = true;
+  services.xserver.desktopManager.gnome.enable = true;
+
+  system.stateVersion = "22.05";
+}
+```
+
+In this tutorial we focus on testing NixOS configurations on a virtual machine, in these cases you can remove the `hardware-configuration.nix` import.
+
+Changes to the configuration need to be positioned inside the curly bracket.[^bracket]
+
+To be able to log in, you must uncomment the section that specifies the user `alice`, or add the following lines:
+```nix
+  users.users.alice = {
+    isNormalUser = true;
+    extraGroups = [ "wheel" ];
+    packages = with pkgs; [
+      firefox
+      thunderbird
+    ];
+  };
+```
+
+Additionally, you need to specify a password for this user.
+An insecure[^password] solution is to specify an plain text password by adding the `initialPassword` option to the user configuration:
+```nix
+    initialPassword = "testpw";
+```
+
+The complete `configuration.nix` file now looks like this:
+```nix
+{ config, pkgs, ... }:
+{
+  boot.loader.systemd-boot.enable = true;
+  boot.loader.efi.canTouchEfiVariables = true;
+
+  services.xserver.enable = true;
+
+  services.xserver.displayManager.gdm.enable = true;
+  services.xserver.desktopManager.gnome.enable = true;
+
+  users.users.alice = {
+    isNormalUser = true;
+    extraGroups = [ "wheel" ]; # Enable ‘sudo’ for the user.
+    packages = with pkgs; [
+      firefox
+      thunderbird
+    ];
+    initialPassword = "testpw";
+  };
+
+  system.stateVersion = "22.05";
+}
+```
+
+## Creating a QEMU based virtual machine using a configuration
+
+A virtual machine is created with the `nix-build` command.
+
+To select `configuration.nix` in the working directory, specify the configuration file as an argument:
+```shell-session
+$ nix-build '<nixpkgs/nixos>' -A vm -I nixos-config=./configuration.nix
+```
+[^nixosrebuild]
+
+This command builds the attribute `vm` utilizing the version of Nixpkgs as specified in the environment variable `NIX_PATH` and using the NixOS configuration as specified in the relative path.
```

Comment:
I'm on the fence about `NIX_PATH`. I'd like to avoid it entirely, but it of course makes the examples super lengthy.

@domenkozar @infinisil do you also prefer to specify the commit hash? Related: https://github.com/NixOS/nix.dev/issues/336

Reply:
> `nix-build '<nixpkgs/nixos>' -A vm -I nixos-config=./configuration.nix` 

how to do this line without <nixpkgs> reference to `NIX_PATH`?

Reply:
I'll be damned, there is indeed no straightforward way. There is a hacky but pure way, but I doubt we should inflict this on anyone.

```console
nix-build <(nix-prefetch-url --print-path https://github.com/NixOS/nixpkgs/tree/78342239aac6382af25986219ffb163ba79b155a)/nixos -A vm -I nixos-config=./configuration.nix
```

Reply:
A simpler way is
```
nix-build '<nixpkgs/nixos>' -A vm \
  -I nixos-config=./configuration.nix \
  -I nixpkgs=https://github.com/NixOS/nixpkgs/tarball/78342239aac6382af25986219ffb163ba79b155a
```

Or even this impure version:

```
nix-build '<nixpkgs/nixos>' -A vm \
  -I nixos-config=./configuration.nix \
  -I nixpkgs=channel:nixos-unstable
```

This fetches the channel from https://channels.nixos.org/

Reply:
@infinisil 

> -I nixpkgs=channel:nixos-unstable

is this documented somewhere?